src/chrome/internal/breakpoints/features/bpRecipeAtLoadedSourceLogic.ts

```diff
@@ -0,0 +1,141 @@
+    private readonly _publishDebuggeeBPRecipeRemoved = this._breakpointsEventsPublisher.publisherForDebuggeeBPRecipeRemoved();
+    private readonly _publishBPRecipeIsUnbound = this._breakpointsEventsPublisher.publisherForBPRecipeIsUnbound();
+
+    public readonly withLogging = wrapWithMethodLogger(this);
+
+    constructor(
+        private readonly _breakpointsEventsPublisher: IBreakpointsEventsPublisher,
+        private readonly _breakpointFeaturesSupport: IBreakpointFeaturesSupport,
+        private readonly _bpRecipesRegistry: DebuggeeBPRsSetForClientBPRFinder,
+        private readonly _targetBreakpoints: IDebuggeeBreakpointsSetter,
+        private readonly _eventsToClientReporter: IEventsToClientReporter,
+        private readonly _debuggeePausedHandler: IDebuggeePausedHandler) {
+        this.doesTargetSupportColumnBreakpointsCached = this._breakpointFeaturesSupport.supportsColumnBreakpoints;
+        this._debuggeePausedHandler.registerActionProvider(paused => this.withLogging.onProvideActionForWhenPaused(paused));
+    }
+
+    public async onProvideActionForWhenPaused(paused: PausedEvent): Promise<IActionToTakeWhenPaused> {
+        if (paused.hitBreakpoints && paused.hitBreakpoints.length > 0) {
+            // TODO DIEGO: Improve this to consider breakpoints where we shouldn't pause
+            return new HitBreakpoint(this._eventsToClientReporter);
+        } else {
+            return new NoActionIsNeededForThisPause(this);
+        }
+    }
+
+    public async addBreakpointAtLoadedSource(bpRecipe: BPRecipeInLoadedSource<ConditionalPause | AlwaysPause>): Promise<CDTPBreakpoint[]> {
+        try {
+            const bpsInScriptRecipe = bpRecipe.mappedToScript();
+
+            const breakpoints = _.flatten(await asyncMap(bpsInScriptRecipe, async bpInScriptRecipe => {
+                const bestLocation = await this.considerColumnAndSelectBestBPLocation(bpInScriptRecipe.location);
+                const bpRecipeInBestLocation = bpInScriptRecipe.withLocationReplaced(bestLocation);
+
+                const runtimeSource = bpInScriptRecipe.location.script.runtimeSource;
+
+                let breakpoints: CDTPBreakpoint[];
+                if (!runtimeSource.doesScriptHasUrl()) {
+                    breakpoints = [await this._targetBreakpoints.setBreakpoint(bpRecipeInBestLocation)];
+                } else {
+                    /**
+                     * If the script is a local file path, we *need* to transform it into an url to be able to set the breakpoint
+                     *
+                     * If the script has an URL and it's not a local file path, then we could actually leave it as-is.
+                     * We transform it into a regexp anyway to add a GUID to it, so CDTP will let us add the same breakpoint/recipe two times (using different guids).
+                     * That way we can always add the new breakpoints for a file, before removing the old ones (except if the script doesn't have an URL)
```

Comment:
I thought that with the breakpoint diffing, we wouldn't have to remove then immediately re-add new BPs?

Reply:
We have several cases. For "normal" breakpoints that are set in a script which has a url we are adding all the new breakpoints first, and removing all the breakpoints that we don't need any more second. We can only do that, because we are adding the random guid to each breakpoint.

If we didn't add a random guid to each breakpoint, we wouldn't be able to add new breakpoint first, and remove old breakpoints later because Chrome considers a breakpoint with and without a condition to be the same breakpoint. If you go to VS Code, and you add a breakpoint, and then you right click to add a condition to it, you are swapping a breakpoint for a conditional breakpoint in that same line. If you tried to add that conditional breakpoint in chrome without using a guid, i'd complain that a breakpoint on that line already exists, so your only option would be to remove the normal breakpoint first, and then add the conditional breakpoint, thus we'd have a race condition. Because we are using the random guid for the breakpoints, the new conditional breakpoint is seen as a different breakpoint than the normal breakpoint, and we can add it before removing the old normal breakpoint.

For scripts without an url, we cannot use a regexp. We cannot add new breakpoints first, and remove old breakpoints later because if a user adds a breakpoint, and then converts it  to a conditional breakpoint, chrome will give us an error, so in that scenario we need to remove the breakpoints first, and add breakpoints later.

The other reason we need the guid, is for a part of the breakpoints logic that we haven't implemented yet. We need to add a component to BreakpointsUpdater.updateBreakpointsForFile on line 100 so when we cannot resolve a source because it's associated script hasn't been loaded, we should set the breakpoint in the project file path to see if it works e.g.: C:\myproject\myap\app.js. That breakpoint will only work for Node.js applications and/or static web applications hosted on storage for Chrome. In those cases this new component that we'll add will add a breakpoint to C:\myproject\myap\app.js, and after the script is loaded, ExistingBPsForJustParsedScriptSetter and BPRecipeAtLoadedSourceLogic will try to re-add the same breakpoint. Having the guid simplifies that logic, because we can just add the breakpoint twice (once by each component), and then let the new component that sets breakpoints before loading remove that breakpoint.

Reply:
What about the diffing though? I must be missing something because the response doesn't seem to answer my question.

> For scripts without an url, we cannot use a regexp

Why not, aren't we doing that currently?